File: h11/h11.m
clear all; clc; close all
% Parâmetros
n_bits = 1000;                % Número de bits
T = 500;                      % Tempo de símbolo OFDM
Ts = 2;                       % Tempo de símbolo em portadora única
K = T/Ts;                     % Número de subportadoras independentes
N = 2*K;                      % DFT de N pontos
sigmas=[0 0.1 1];             % Vetor de variâncias do ruído
%
% Gerar bits aleatórios
dataIn=rand(1,n_bits);   % Sequência de números entre 0 e 1 uniformemente distribuídos
dataIn=sign(dataIn-.5);  % Sequência de -1 e 1
% Conversor serial paralelo
dataInMatrix = reshape(dataIn,n_bits/4,4);
%
% Gerar constelaçao 16-QAM
seq16qam = 2*dataInMatrix(:,1)+dataInMatrix(:,2)+1i*(2*dataInMatrix(:,3)+dataInMatrix(:,4)); 
seq16=seq16qam';
% Garantir propriedadade da simetria
X = [seq16 conj(seq16(end:-1:1))]; 
%
% Construindo xn
xn = zeros(1,N);
for n=0:N-1
    for k=0:N-1
        xn(n+1) = xn(n+1) + 1/sqrt(N)*X(k+1)*exp(1i*2*pi*n*k/N);
    end
end
% 
% Loop de variâncias
for ik = 1:length(sigmas)
    %
    % Adição de ruído
    variance = sigmas(ik);
    noise = sqrt(variance)*randn(1,N)+1i*sqrt(variance)*randn(1,N);
    %
    % sinal recebido = xn + ruído 
    rn = xn+noise;
    % DFT de rn
    Y = zeros(1,K);
    for k=0:K-1
        for n=0:N-1
            Y(1,k+1) = Y(1,k+1) + 1/sqrt(N)*rn(n+1)*exp(-1i*2*pi*k*n/N);
        end
    end
    %
    % Plots
    scatterplot(Y)
    hold on
    scatter(real(seq16),imag(seq16), 'r', '+')
    hold off
    title(['Sinal com ruído de variância ', num2str(variance)]);
    % Demodulação  
    for k= 1:length(Y) % Para percorrer todo o vetor Yk 
        if real(Y(1,k)) > 0 % Para parte real de Yk positiva
            if real(Y(1,k)) > 2
                Z(1,k) = 3;
            else
                Z(1,k) = 1;
            end
        else % Para parte real de Yk negativa ou igual a zero
            if real(Y(1,k)) < -2
                 Z(1,k) = -3;
            else
                 Z(1,k) = -1;
            end
        end

        if imag(Y(1,k)) > 0 % Para parte imaginaria de Yk positiva
            if imag(Y(1,k)) > 2
                Z(1,k) = Z(1,k) + 1i*3;
            else
                Z(1,k) = Z(1,k) + 1i;
            end
        else % Para parte imaginaria de Yk negativa ou igual a zero
            if imag(Y(1,k)) < -2
                 Z(1,k) = Z(1,k) - 1i*3;
            else
                 Z(1,k) = Z(1,k) - 1i;
            end
        end
    end
    % Contagem de erro
    error = length(find(Z(1,2:K)-X(1,2:K)));
    disp(['Para variância de ', num2str(variance), ' houve ', num2str(error), ' símbolos errados.']);
end

File: h11/h11.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.special import erfc

# Parâmetros
n_bits = int(1000)          # Número de bits
T = int(500)                # Tempo de símbolo OFDM
Ts = 2                      # Tempo de símbolo em portadora única
K = int(T/Ts)               # Número de subportadoras independentes
N = int(2*K)                # DFT de N pontos
sigmas=[0, 0.1, 1]          # Vetor de variâncias do ruído
vtEbN0_dB = np.arange(0,15) 

# Gerar bits aleatórios
dataIn=np.random.rand(n_bits)   # Sequência de números entre 0 e 1 uniformemente distribuídos
dataIn=np.sign(dataIn-.5)       # Sequência de -1 e 1
# Conversor serial paralelo
dataInMatrix = dataIn.reshape((int(n_bits/4),4))

# Gerar constelaçao 16-QAM
seq16 = 2*dataInMatrix[:,0]+dataInMatrix[:,1]+1j*(2*dataInMatrix[:,2]+dataInMatrix[:,3])

# Garantir propriedadade da simetria
X = np.concatenate([seq16, np.conj(seq16[::-1])])

# Construindo xn
xn = np.zeros((N), dtype=complex)
xn = (N/np.sqrt(N))*np.fft.ifft(X)

# Loop de variâncias
error = np.zeros(len(vtEbN0_dB))

for ik in range(len(vtEbN0_dB)):
    
    # Adição de ruído
    # sinal recebido = xn + ruído 
    noise = 1/np.sqrt(2)*(np.random.randn(int(N))+1j*np.random.randn(int(N)))
    noise = noise*10**(-vtEbN0_dB[ik]/10)
    rn = xn+noise
    # DFT de rn
    Y = np.zeros((K), dtype=complex)
    Y = (1/np.sqrt(N))*np.fft.fft(rn)
    
    # Demodulação 
    Z = np.zeros(len(Y), dtype=complex)
    for k in range(len(Y)): # Para percorrer todo o vetor Yk 
        if np.real(Y[k]) > 0: # Para parte real de Yk positiva
            if np.real(Y[k]) > 2:
                Z[k] = 3
            else:
                Z[k] = 1
        else: # Para parte real de Yk negativa ou igual a zero
            if np.real(Y[k]) < -2:
                 Z[k] = -3
            else:
                 Z[k] = -1

        if np.imag(Y[k]) > 0: # Para parte imaginaria de Yk positiva
            if np.imag(Y[k]) > 2:
                Z[k] = Z[k] + 1j*3
            else:
                Z[k] = Z[k] + 1j

        else: # Para parte imaginaria de Yk negativa ou igual a zero
            if np.imag(Y[k]) < -2:
                 Z[k] = Z[k] - 1j*3
            else:
                 Z[k] = Z[k] - 1j

    # Contagem de erro
    error= len(np.where(Z-X!=0)[0])
    print("Para EbN0 = " + str(vtEbN0_dB[ik])  + " dB, ha " +str(error) +" simbolos errados" )